### PR TITLE
OpenZFS 9403 - assertion failed in arc_buf_destroy() when concurrentl…

### PR DESCRIPTION
**Pull request by tcaputi**
…y reading block with checksum error

Porting notes:
* zio_decompress_fail_fraction was made into a tunable. In the
  future this could be added to zinject.

Authored by: Matt Ahrens <mahrens@delphix.com>
Reviewed by: George Wilson <george.wilson@delphix.com>
Reviewed by: Paul Dagnelie <pcd@delphix.com>
Reviewed by: Pavel Zakharov <pavel.zakharov@delphix.com>
Approved by: Matt Ahrens <mahrens@delphix.com>
Ported-by: Tom Caputi <tcaputi@datto.com>

OpenZFS-issue: https://illumos.org/issues/9403
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/fa98e487a9

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: module/zfs/arc.c [hunk 1]
```diff
@@ -5803,10 +5803,11 @@ arc_getbuf_func(zio_t *zio, const zbookmark_phys_t *zb, const blkptr_t *bp,
 	arc_buf_t **bufp = arg;
 
 	if (buf == NULL) {
+		ASSERT(zio == NULL || zio->io_error != 0);
 		*bufp = NULL;
 	} else {
```
Comment (behlendorf) [inline module/zfs/arc.c]: It looks like you accidentally dropped an ASSERT from the OpenZFS version which should be here.

```
ASSERT(zio == NULL || zio->io_error == 0);
```

There are also two additional ASSERTs which were lost and should be added to `dbuf_prefetch_indirect_done()`.

File: module/zfs/zio_compress.c [hunk 2]
```diff
@@ -148,5 +154,23 @@ zio_decompress_data(enum zio_compress c, abd_t *src, void *dst,
...
 
+	/*
+	 * Decompression shouldn't fail, because we've already verifyied
+	 * the checksum.  However, for extra protection (e.g. against bitflips
+	 * in non-ECC RAM), we handle this error (and test it).
+	 */
+	ASSERT0(ret);
+	if (zio_decompress_fail_fraction != 0 &&
+	    spa_get_random(zio_decompress_fail_fraction) == 0)
+		ret = SET_ERROR(EINVAL);
```
Comment (behlendorf) [inline module/zfs/zio_compress.c]: After taking a second look at this I think it's worth doing this decompress error injection with `zinject -e decompress`.  You can model it after the existing `zinject -e decrypt` support and the `zio_handle_decrypt_injection` handler used in `zio_decrypt`. 

This makes the code consistent and allows for an easy test case by adding a new `fault/decompress_fault.ksh` based on the existing `fault/decrypt_fault.ksh`.

File: module/zfs/zio.c [hunk 3]
```diff
@@ -389,6 +389,12 @@ zio_decompress(zio_t *zio, abd_t *data, uint64_t size)
 		    zio->io_abd, tmp, zio->io_size, size);
 		abd_return_buf_copy(data, tmp, size);
 
+#ifdef _KERNEL
+		printk(KERN_DEBUG "HERE1: %d %d\n", zio_injection_enabled, ret);
+#endif
```
Comment (behlendorf) [inline module/zfs/zio.c]: Stray debug line needs to be removed.

File: tests/zfs-tests/tests/functional/fault/decompress_fault.ksh [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+{
+	log_must eval "echo 1 > /sys/module/zfs/parameters/zfs_compressed_arc_enabled"
+	log_must zinject -c all
+	default_cleanup_noexit
+}
+
+log_onexit cleanup
+
+default_mirror_setup_noexit $DISK1 $DISK2
+log_must eval "echo 0 > /sys/module/zfs/parameters/zfs_compressed_arc_enabled"
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/fault/decompress_fault.ksh]: Use `log_must set_tunable64 zfs_compressed_arc_enabled 0` for this and the cleanup.

File: module/zfs/zio.c (hunk 3 above)
Comment (tcaputi) [reply module/zfs/zio.c]: Sorry about that.

File: tests/zfs-tests/tests/functional/fault/decompress_fault.ksh (hunk 4 above)
Comment (tcaputi) [reply tests/zfs-tests/tests/functional/fault/decompress_fault.ksh]: Thanks for letting me know that that existed. I will make sure to use it in the future.

**Pull request merged**